### PR TITLE
:sparkles: Add openstack kamaji cluster stack

### LINKED ISSUE #25: Clusterclass with kamaji
My very first steps/impression: The controller to create so called tenantcontrolplanes (kamajis core-functionality) works so far pretty nicely. The next layer is to create such a tenantcontrolplane via a cluter-API controlplane (kamajicontrolplane) resource. So far i have not been successful in doing that. Observation:  when a kamajicontrolplane resource is created alongside with a supported provider (for example hcloud). The resource is taken care of, otherwise the controller ignores it.

### PR DESCRIPTION
**Pull request by chess-knight**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #25

**Special notes for your reviewer**:
You can use [openstack-kamaji-1-30-v0-sha-11930ee](https://github.com/SovereignCloudStack/cluster-stacks/releases/tag/openstack-kamaji-1-30-v0-sha-11930ee) release for the tests.
```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: kamaji-cluster
  labels:
    managed-secret: cloud-config
spec:
  clusterNetwork:
    services:
      cidrBlocks: ["10.128.0.0/12"]
    pods:
      cidrBlocks: ["192.168.0.0/16"]
    serviceDomain: "cluster.local"
  topology:
    class: openstack-kamaji-1-30-v0-sha.11930ee
    controlPlane:
      replicas: 3
    variables:
    - name: data_store
      value: "default"
    - name: dns_service_ips
      value: ["10.128.0.10"]
    - name: worker_flavor
      value: "SCS-2V-4-50"
    - name: external_id
      value: "ebfe5546-f09f-4f42-ab54-094e457d42ec"
    version: v1.30.1
    workers:
      machineDeployments:
      - class: default-worker
        name: kamaji
        replicas: 3
        failureDomain: nova

```

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squash commits
- [ ] include documentation
- [ ] add unit tests

### CODE REVIEW

**Review by paulphys — APPROVED** [review]

File: providers/openstack/kamaji/1-30/cluster-class/templates/kamaji-control-plane-template.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+apiVersion: controlplane.cluster.x-k8s.io/v1alpha1
+kind: KamajiControlPlaneTemplate
+metadata:
+  name: {{ .Release.Name }}-{{ .Chart.Version }}-control-plane
+spec:
+  template:
+    spec:
```
Comment (michal-gubricky) [inline providers/openstack/kamaji/1-30/cluster-class/templates/kamaji-control-plane-template.yaml:7]: The field `dataStoreName` is probably missing here, attaching cso logs
```
{"level":"ERROR","time":"2024-06-17T08:01:21.390Z","file":"kube/kube.go:107","message":"failed to apply object","controller":"clusterstackrelease","controllerGroup":"clusterstack.x-k8s.io","controllerKind":"ClusterStackRelease","ClusterStackRelease":{"name":"openstack-kamaji-1-30-v0-sha-11930ee","namespace":"default"},"namespace":"default","name":"openstack-kamaji-1-30-v0-sha-11930ee","reconcileID":"59c2788e-1bb8-4934-92cd-c297d26cc72c","obj":"controlplane.cluster.x-k8s.io/v1alpha1, Kind=KamajiControlPlaneTemplate","name":"openstack-kamaji-1-30-v0-sha.11930ee-control-plane","namespace":"","error":"failed to apply object: KamajiControlPlaneTemplate.controlplane.cluster.x-k8s.io \"openstack-kamaji-1-30-v0-sha.11930ee-control-plane\" is invalid: spec.template.spec.dataStoreName: Required value","stacktrace":"github.com/SovereignCloudStack/cluster-stack-operator/pkg/kube.(*kube).Apply\n\t/src/cluster-stack-operator/pkg/kube/kube.go:107\ngithub.com/SovereignCloudStack/cluster-stack-operator/internal/controller.(*ClusterStackReleaseReconciler).templateAndApply\n\t/src/cluster-stack-operator/internal/controller/clusterstackrelease_controller.go:286\ngithub.com/SovereignCloudStack/cluster-stack-operator/internal/controller.(*ClusterStackReleaseReconciler).Reconcile\n\t/src/cluster-stack-operator/internal/controller/clusterstackrelease_controller.go:169\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/src/cluster-stack-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:119\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/src/cluster-stack-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:316\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/src/cluster-stack-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/src/cluster-stack-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227"}
```

Comment (chess-knight) [reply providers/openstack/kamaji/1-30/cluster-class/templates/kamaji-control-plane-template.yaml:7]: Thank you, I will look into it!

Comment (chess-knight) [reply providers/openstack/kamaji/1-30/cluster-class/templates/kamaji-control-plane-template.yaml:7]: Please, try it with `control-plane-components.yaml` from https://github.com/clastix/cluster-api-control-plane-provider-kamaji/pull/111

Comment (michal-gubricky) [reply providers/openstack/kamaji/1-30/cluster-class/templates/kamaji-control-plane-template.yaml:7]: Cool, I will try and let you know

Comment (michal-gubricky) [reply providers/openstack/kamaji/1-30/cluster-class/templates/kamaji-control-plane-template.yaml:7]: After applying the new CRDs, the clusterstackrelease is now ready
```
$ k get clusterstackreleases.clusterstack.x-k8s.io 
NAME                                   K8S VERSION   READY   AGE     REASON   MESSAGE
openstack-kamaji-1-30-v0-sha-11930ee   v1.30.1       true    5h20m   
```

**Comment by chess-knight** [comment]
> After applying the cluster resource I got the following error:
> 
> ```
> message: 'cannot create or update TenantControlPlane: admission webhook vtenantcontrolplane.kb.io" denied the request: unable to parse the desired Kubernetes version: Invalid character(s) found in major number "v1"'
> ```

This is already fixed(but not released yet) in the upstream, see https://github.com/clastix/cluster-api-control-plane-provider-kamaji/pull/109. You can workaround that by editing the `KamajiControlPlane` resource and removing "v" prefix from `.spec.version` field.

**Comment by mxmxchere** [comment]
I have the feeling that we still lack `KamajiControlPlaneTemplate` is there a release newer than [v0.9.0](https://github.com/clastix/cluster-api-control-plane-provider-kamaji/releases/tag/v0.9.0)?

**Comment by chess-knight** [comment]
> I have the feeling that we still lack `KamajiControlPlaneTemplate`. Is there a release newer than [v0.9.0](https://github.com/clastix/cluster-api-control-plane-provider-kamaji/releases/tag/v0.9.0)?

Yes, there is no upstream release yet, please see https://github.com/SovereignCloudStack/cluster-stacks/pull/104#discussion_r1642770311.

**Comment by jschoone** [comment]
[v0.10.0](https://github.com/clastix/cluster-api-control-plane-provider-kamaji/releases/tag/v0.10.0) is now available :)

**Comment by mxmxchere** [comment]
I am testing with 1.0.0 and capk 0.10.0:
`tenantcontrolplane.spec.addons.konnectivity` (kamaji-resource) has a required field `proxyPort` but 
`kamajicontrolplane.spec.addons.konnectivity` (capi-wrapper resource) does not know this field

**Comment by chess-knight** [comment]
> I am testing with kamaji 1.0.0 and capk 0.10.0: `tenantcontrolplane.spec.addons.konnectivity` (kamaji-resource) has a required field `proxyPort` but `kamajicontrolplane.spec.addons.konnectivity` (capi-wrapper resource) does not know this field

This is strange. Does it work with KAmaji v0.6x?

**Review by mxmxchere — APPROVED** [review]

**Pull request merged**